File: 3493.properties-graph.java
#
# @lc app=leetcode id=3493 lang=java
#
# [3493] Properties Graph
#

# @lc code=start
class Solution {
    private int find(int x, int[] parent) {
        int root = x;
        while (parent[root] != root) {
            root = parent[root];
        }
        int cur = x;
        while (cur != root) {
            int nxt = parent[cur];
            parent[cur] = root;
            cur = nxt;
        }
        return root;
    }

    public int numberOfComponents(int[][] properties, int k) {
        int n = properties.length;
        boolean[][] pres = new boolean[n][101];
        for (int i = 0; i < n; i++) {
            for (int val : properties[i]) {
                pres[i][val] = true;
            }
        }
        int[] parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int common = 0;
                for (int v = 1; v <= 100; v++) {
                    if (pres[i][v] && pres[j][v]) {
                        common++;
                    }
                }
                if (common >= k) {
                    int pi = find(i, parent);
                    int pj = find(j, parent);
                    if (pi != pj) {
                        parent[pi] = pj;
                    }
                }
            }
        }
        int comp = 0;
        for (int i = 0; i < n; i++) {
            if (find(i, parent) == i) {
                comp++;
            }
        }
        return comp;
    }
}
# @lc code=end

File: 3493.properties-graph.go
#
# @lc app=leetcode id=3493 lang=golang
#
# [3493] Properties Graph
#
# @lc code=start
func numberOfComponents(properties [][]int, k int) int {
    n := len(properties)
    
    // Helper function to calculate intersection count of distinct integers
    intersect := func(a, b []int) int {
        setA := make(map[int]bool)
        for _, val := range a {
            setA[val] = true
        }
        
        count := 0
        counted := make(map[int]bool)
        for _, val := range b {
            if setA[val] && !counted[val] {
                count++
                counted[val] = true
            }
        }
        return count
    }
    
    // Union-Find data structure
    parent := make([]int, n)
    for i := range parent {
        parent[i] = i
    }
    
    var find func(int) int
    find = func(x int) int {
        if parent[x] != x {
            parent[x] = find(parent[x]) // Path compression
        }
        return parent[x]
    }
    
    union := func(x, y int) {
        rootX := find(x)
        rootY := find(y)
        if rootX != rootY {
            parent[rootX] = rootY
        }
    }
    
    // Build the graph by checking all pairs
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            if intersect(properties[i], properties[j]) >= k {
                union(i, j)
            }
        }
    }
    
    // Count connected components
    components := make(map[int]bool)
    for i := 0; i < n; i++ {
        components[find(i)] = true
    }
    
    return len(components)
}
# @lc code=end